lib/recursive-methods.rb

```diff
@@ -1,49 +1,75 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), it will run the number of n times
+# Space complexity: O(n), it will call the function the number of n times
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n <= 1
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length of s
+# Space complexity: O(n^2), where n is the length of s. Will call the function and create a new string n number of times, so n^2
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1
+    return s[-1] + reverse(s[0..-2])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - n is length of s divided by 2 since it stops once last_index and first_index meet in the middle
+# Space complexity: O(n) - same as time complexity, but does not create a new string each time so doesn't become O(n^2)
 def reverse_inplace(s)
```

Comment:
👍 This works, but why send the 1st character to the helper method?  Why not just make a local temp variable?